libbeat/cmd/instance/locks/lock.go

```diff
@@ -0,0 +1,215 @@
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path: %w", err)
+	}
+
+	err = os.Remove(lock.fileLock.Path())
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path: %w", err)
+	}
+	return nil
+}
+
+// ******* private helpers
+
+// createPidfile creates the underling pidfile used by the lock
+// This assumes that no previous lockfile should exist.
+// returns a bool indicating if the file was created, and an error
+func (lock *Locker) createPidfile(pid int) (bool, error) {
+	_, err := os.Stat(lock.filePath)
+	//file already exists, go into repair/handle fail mode
+	if !os.IsNotExist(err) {
+		return false, nil
+	}
+
+	new := pidfile{Pid: pid, WriteTime: time.Now()}
+	encoded, err := json.Marshal(&new)
+	if err != nil {
+		return false, fmt.Errorf("error encoding json for pidfile: %w", err)
+	}
+
+	err = os.WriteFile(lock.filePath, encoded, os.FileMode(0600))
+	if err != nil {
+		return false, fmt.Errorf("error writing pidfile to %s: %w", lock.filePath, err)
+	}
+	return true, nil
+}
+
+// handleFailedLock will attempt to recover from a failed lock operation in a pid-aware way.
+// The point of this is to deal with instances where an improper beat shutdown left us with
+// a pre-existing pidfile for a beat process that no longer exists.
+// The one argument tells the method if a pre-existing lockfile was already found.
+func (lock *Locker) handleFailedLock() error {
+	// read in whatever existing lockfile caused us to fail
+	pf, err := lock.readExistingPidfile()
+	if err != nil {
+		return fmt.Errorf("error reading existing lockfile: %w", err)
+	}
+
+	// Case: the lockfile is locked, but by us. Probably a coding error,
+	// and probably hard to do
+	if pf.Pid == os.Getpid() {
+		return fmt.Errorf("lockfile for beat has been locked twice by the same PID, potential bug.")
+	}
+
+	// Check to see if the PID found in the pidfile exists.
+	// Checking if the pid exists in a cross-platform way is more fraught than it looks.
+	// we have internal metrics libraries we can use for this, but all those will use APIs
+	// dedicated to gathering extended process info and metrics, which can come with extra permissions hurdles,
+	// making those methods more likely to return an error.
+	exists, err := process.PidExistsWithContext(context.Background(), int32(pf.Pid))
+	if err != nil {
+		return fmt.Errorf("error looking up status for pid %d: %w", pf.Pid, err)
+	}
+
+	// Case: we've gotten a lock file for another process that's already running
+	// This is the "base" lockfile case, which is two beats running from the same directory
+	if exists {
+		// This will make debugging easier for someone.
+		state, err := metricproc.GetInfoForPid(resolve.NewTestResolver("/"), pf.Pid)
+		// Above call is is auxiliary debug data, so we don't care too much if it fails
+		debugString := fmt.Sprintf("process with PID %d", pf.Pid)
+		if err == nil {
+			debugString = fmt.Sprintf("process '%s' with pid %d", state.Name, pf.Pid)
```

Comment:
```suggestion
			debugString = fmt.Sprintf("process '%s' with PID %d", state.Name, pf.Pid)
```
nit: let's be consistent when formatting the log messages